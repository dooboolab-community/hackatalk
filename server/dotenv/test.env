JWT_SECRET=dooboolab
DATABASE_URL="postgresql://test:test!@localhost:5432/postgres?schema=test"
