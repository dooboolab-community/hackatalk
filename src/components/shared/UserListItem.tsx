import { ImageSourcePropType, TouchableOpacity, View, ViewStyle } from 'react-native';

import CheckBox from './CheckBox';
import { Ionicons } from '@expo/vector-icons';
import React from 'react';
import { User } from '../../types';
import styled from 'styled-components/native';
import { useThemeContext } from '@dooboo-ui/native-theme';

interface Props {
  testID?: string;
  style?: ViewStyle;
  user: User;
  onPress?: () => void;
  onLongPress?: () => void;
  showCheckBox?: boolean;
  checked?: boolean;
}

const Container = styled.View`
  width: 100%;
`;

const Wrapper = styled.View`
  background-color: ${({ theme }): string => theme.itemBackground};
  height: 80px;
  border-bottom-width: 1px;
  border-color: ${({ theme }): string => theme.backgroundDark};
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
  padding: 0 20px;
`;

const StyledImage = styled.Image`
  width: 40px;
  height: 40px;
`;

const StyledText = styled.Text`
  margin-left: 12px;
  width: 200px;
  font-size: 14px;
  color: ${({ theme }): string => theme.fontColor};
`;

const StyledRightText = styled.Text`
  position: absolute;
  right: 20px;
  font-size: 12px;
  color: ${({ theme }): string => theme.fontSubColor};
  max-width: 120px;
  border-width: 0.3px;
  border-color: ${({ theme }): string => theme.lineColor};
  padding: 4px 8px;
`;

function Shared({
  showCheckBox = false,
  checked = false,
  onPress,
  onLongPress,
  testID,
  user: { photoURL = '', nickname, statusMessage },
}: Props): React.ReactElement {
  const { theme } = useThemeContext();
  const photoURLObj: ImageSourcePropType =
    typeof photoURL === 'string' ? { uri: photoURL } : photoURL;

  return (
    <Container>
      <TouchableOpacity
        testID={testID}
        activeOpacity={0.5}
        onPress={onPress}
        onLongPress={onLongPress}
      >
        <Wrapper>
          {
            photoURL
              ? <StyledImage source={photoURLObj} />
              : <View
                style={{
                  width: 40,
                  height: 40,
                  alignItems: 'center',
                  justifyContent: 'center',
                }}
              >
                <Ionicons name="ios-person" size={24} color={theme.fontColor} />
              </View>
          }
          <StyledText
            numberOfLines={1}
          >{nickname}</StyledText>
          {
            showCheckBox
              ? <CheckBox
                containerStyle={{
                  position: 'absolute',
                  right: 20,
                }}
                checkColor={theme.tintColor}
                hasChecked={checked}
                onToggle={onPress}
              />
              : statusMessage
                ? <StyledRightText>{statusMessage}</StyledRightText>
                : null
          }
        </Wrapper>
      </TouchableOpacity>
    </Container>
  );
}

Shared.defaultProps = {
  user: {
    id: '',
    nickname: '',
    photoURL: null,
    statusMessage: '',
    isOnline: false,
    friends: [],
    channels: [],
    created: undefined,
    updated: undefined,
  },
};
export default Shared;
