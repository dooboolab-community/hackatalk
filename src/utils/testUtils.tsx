import React, { ReactElement } from 'react';

import RootProvider from '../providers';
import { ThemeType } from '../types';

export const createTestElement = (
  child: ReactElement,
  themeType?: ThemeType,
): ReactElement => (
  <RootProvider initialThemeType={themeType}>{child}</RootProvider>
);

export const createTestProps = (
  obj: object = {},
  moreScreenProps?: object,
): object | unknown | any => ({
  navigation: {
    navigate: jest.fn(),
    goBack: jest.fn(),
    resetRoot: jest.fn(),
  },
  screenProps: {
    changeThemeType: jest.fn(),
    ...moreScreenProps,
  },
  ...obj,
});
