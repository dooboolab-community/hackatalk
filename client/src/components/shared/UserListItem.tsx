import { ImageSourcePropType, TouchableOpacity, View, ViewStyle } from 'react-native';

import CheckBox from './CheckBox';
import { IC_NO_IMAGE } from '../../utils/Icons';
import React from 'react';
import { User } from '../../types/graphql';
import styled from 'styled-components/native';
import { useThemeContext } from '@dooboo-ui/theme';

interface Props {
  testID?: string;
  style?: ViewStyle;
  user: User;
  onPress?: () => void;
  onLongPress?: () => void;
  showCheckBox?: boolean;
  showStatus?: boolean;
  checked?: boolean;
}

const Container = styled.View`
  width: 100%;
`;

const Wrapper = styled.View`
  background-color: ${({ theme }): string => theme.itemBackground};
  height: 80px;
  border-bottom-width: 1px;
  border-color: ${({ theme }): string => theme.backgroundDark};
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
  padding: 0 20px;
`;

const ImageWrapper = styled.View`
  width: 40px;
  height: 40px;
  align-items: center;
  justify-content: center;
`;

const StatusTag = styled.View`
  position: absolute;
  top: 0;
  right: 0;
  width: 12px;
  height: 12px;
  border-radius: 25px;
`;

const StyledImage = styled.Image`
  width: 40px;
  height: 40px;
  border-radius: 20px;
`;

const StyledText = styled.Text`
  margin-left: 12px;
  width: 200px;
  font-size: 14px;
  font-weight: bold;
  color: ${({ theme }): string => theme.fontColor};
`;

const StyledRightText = styled.Text`
  position: absolute;
  right: 20px;
  font-size: 12px;
  color: ${({ theme }): string => theme.fontSubColor};
  max-width: 120px;
  border-width: 0.3px;
  border-color: ${({ theme }): string => theme.lineColor};
  padding: 4px 8px;
`;

function Shared({
  showCheckBox = false,
  showStatus = false,
  checked = false,
  onPress,
  onLongPress,
  testID,
  user: { photoURL = '', email, nickname, name, statusMessage, isOnline },
}: Props): React.ReactElement {
  const { theme } = useThemeContext();

  const photoURLObj: ImageSourcePropType | null =
    typeof photoURL === 'string'
      ? { uri: photoURL }
      : photoURL;

  return (
    <Container>
      <TouchableOpacity
        testID={testID}
        activeOpacity={0.5}
        onPress={onPress}
        onLongPress={onLongPress}
      >
        <Wrapper>
          <ImageWrapper>
            <StyledImage
              source={
                photoURL && photoURLObj
                  ? photoURLObj
                  : IC_NO_IMAGE
              }
            />
            {
              showStatus
                ? <StatusTag style={{ backgroundColor: isOnline ? '#00D4AB' : '#AFB4C3' }} />
                : null
            }
          </ImageWrapper>
          <StyledText numberOfLines={1}>{name || nickname}</StyledText>
          {
            showCheckBox
              ? <CheckBox
                containerStyle={{
                  position: 'absolute',
                  right: 20,
                }}
                backgroundColor={theme.background}
                checkColor={theme.font}
                inActiveColor={theme.inactiveColor}
                activeColor={theme.checkBackground}
                hasChecked={checked}
                onToggle={onPress}
              />
              : statusMessage
                ? <StyledRightText>{statusMessage}</StyledRightText>
                : null
          }
        </Wrapper>
      </TouchableOpacity>
    </Container>
  );
}

Shared.defaultProps = {
  user: {
    id: '',
    name: '',
    photoURL: null,
    statusMessage: '',
    isOnline: false,
    friends: [],
    channels: [],
    created: undefined,
    updated: undefined,
  },
};

export default Shared;
