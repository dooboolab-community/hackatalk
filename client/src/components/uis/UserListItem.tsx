import {
  ImageSourcePropType,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';
import {graphql, useFragment} from 'react-relay/hooks';

import CheckBox from './CheckBox';
import {FontAwesome} from '@expo/vector-icons';
import {IC_NO_IMAGE} from '../../utils/Icons';
import React from 'react';
import {UserListItem_user$key} from '../../__generated__/UserListItem_user.graphql';
import {getString} from '../../../STRINGS';
import styled from '@emotion/native';
import {useTheme} from 'dooboo-ui';

const fragment = graphql`
  fragment UserListItem_user on User {
    id
    photoURL
    nickname
    name
    statusMessage
    isOnline
    hasBlocked
  }
`;

interface Props {
  style?: ViewStyle;
  user: UserListItem_user$key;
  onPress?: () => void;
  onLongPress?: () => void;
  showCheckBox?: boolean;
  showStatus?: boolean;
  checked?: boolean;
}

const Container = styled.View`
  width: 100%;
`;

const Wrapper = styled.View`
  background-color: ${({theme}) => theme.itemBackground};
  height: 80px;
  border-bottom-width: 1px;
  border-color: ${({theme}) => theme.backgroundDark};
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
  padding: 0 20px;
`;

const ImageWrapper = styled.View`
  width: 40px;
  height: 40px;
  align-items: center;
  justify-content: center;
`;

const StatusTag = styled.View`
  position: absolute;
  top: 0;
  right: 0;
  width: 12px;
  height: 12px;
  border-radius: 25px;
`;

const StyledImage = styled.Image`
  width: 40px;
  height: 40px;
  border-radius: 20px;
`;

const StyledText = styled.Text`
  margin-left: 12px;
  width: 200px;
  font-size: 14px;
  font-weight: bold;
  color: ${({theme}) => theme.text};
`;

const StyledRightText = styled.Text`
  position: absolute;
  right: 20px;
  font-size: 12px;
  color: ${({theme}) => theme.secondaryText};
  max-width: 120px;
  border-width: 0.3px;
  border-color: ${({theme}) => theme.lineColor};
  padding: 4px 8px;
`;

function Shared({
  showCheckBox = false,
  showStatus = false,
  checked = false,
  onPress,
  onLongPress,
  user,
}: Props): React.ReactElement {
  const {
    photoURL = '',
    nickname,
    name,
    statusMessage,
    isOnline,
    hasBlocked,
  } = useFragment(fragment, user);

  const {theme} = useTheme();

  const photoURLObj: ImageSourcePropType | null =
    typeof photoURL === 'string' ? {uri: photoURL} : photoURL;

  return (
    <Container>
      <TouchableOpacity
        testID="peer-button"
        activeOpacity={0.5}
        onPress={onPress}
        onLongPress={onLongPress}>
        <Wrapper>
          <ImageWrapper>
            <StyledImage
              resizeMode="cover"
              source={photoURL && photoURLObj ? photoURLObj : IC_NO_IMAGE}
            />
            {showStatus ? (
              <StatusTag
                style={{backgroundColor: isOnline ? '#00D4AB' : '#AFB4C3'}}
              />
            ) : null}
            {hasBlocked ? (
              <View
                style={{
                  position: 'absolute',
                  top: 0,
                  left: 0,
                }}>
                <FontAwesome
                  name="ban"
                  size={16}
                  color={hasBlocked ? 'red' : 'white'}
                />
              </View>
            ) : null}
          </ImageWrapper>
          <StyledText numberOfLines={1}>
            {nickname || name || getString('NO_NAME')}
          </StyledText>
          {showCheckBox ? (
            <CheckBox
              containerStyle={{
                position: 'absolute',
                right: 20,
              }}
              backgroundColor={theme.background}
              checkColor={theme.text}
              inActiveColor={theme.inactiveColor}
              activeColor={theme.checkBackground}
              hasChecked={checked}
              onToggle={onPress}
            />
          ) : statusMessage ? (
            <StyledRightText>{statusMessage}</StyledRightText>
          ) : null}
        </Wrapper>
      </TouchableOpacity>
    </Container>
  );
}

Shared.defaultProps = {
  user: {
    id: '',
    name: '',
    photoURL: null,
    statusMessage: '',
    isOnline: false,
    friends: [],
    channels: [],
    createdAt: undefined,
    updatedAt: undefined,
  },
};

export default Shared;
