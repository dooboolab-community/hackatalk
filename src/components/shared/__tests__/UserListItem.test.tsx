import 'react-native';

import * as React from 'react';

import {
  RenderResult,
  act,
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react-native';
import { createTestElement, createTestProps } from '../../../../test/testUtils';

import UserListItem from '../UserListItem';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

let cnt = 0;
const onPress = (): void => {
  cnt++;
};

const propsObj = {
  testID: 'test_yo',
  user: {
    uid: '',
    displayName: '',
    thumbURL: null,
    photoURL: null,
    statusMsg: '',
    online: false,
    created: undefined,
    updated: undefined,
  },
  onPress,
};

describe('[UserListItem] rendering test', () => {
  let props;
  let component;

  beforeEach(() => {
    props = createTestProps(propsObj);
    component = createTestElement(<UserListItem {...props} />);
  });

  it('renders as expected', () => {
    const json = renderer.create(component).toJSON();
    expect(json).toMatchSnapshot();
  });
});

describe('[UserListItem] interaction', () => {
  let testingLib: RenderResult;
  let props;
  let component;

  beforeEach(() => {
    props = createTestProps(propsObj);
    component = createTestElement(<UserListItem {...props} />);
    testingLib = render(component);
  });

  afterAll(() => {
    cleanup();
  });

  it('should fireEvent when peer image is pressed', () => {
    act(() => {
      fireEvent.press(testingLib.getByTestId('test_yo'));
    });
    expect(cnt).toEqual(1);
  });
});
