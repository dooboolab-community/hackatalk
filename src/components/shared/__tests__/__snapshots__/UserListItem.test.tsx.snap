// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[UserListItem] rendering test renders as expected 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "width": "100%",
          },
        ]
      }
    >
      <TouchableOpacity
        activeOpacity={0.5}
        testID="test_yo"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffffff",
                "borderBottomWidth": 1,
                "borderColor": "#e2f2ff",
                "flexDirection": "row",
                "height": 80,
                "justifyContent": "flex-start",
                "paddingBottom": 0,
                "paddingLeft": 20,
                "paddingRight": 20,
                "paddingTop": 0,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "width": 40,
              }
            }
          >
            <Text />
          </View>
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#151A25",
                  "fontSize": 14,
                  "marginLeft": 12,
                  "width": 200,
                },
              ]
            }
          >
            
          </Text>
        </View>
      </TouchableOpacity>
    </View>
  </View>
</View>
`;
